lib/ansible/module_utils/postgres.py

```diff
@@ -63,7 +65,50 @@ def postgres_common_argument_spec():
     )
 
 
-def connect_to_db(module, kw, autocommit=False):
+def ensure_required_libs(module):
+    if not HAS_PSYCOPG2:
+        module.fail_json(msg=missing_required_lib('psycopg2'))
+
+    if module.params.get('ca_cert') and psycopg2.__version__ < '2.4.3':
+        module.fail_json(msg='psycopg2 must be at least 2.4.3 in order to user the ca_cert parameter')
+
+
+def connect_to_db(module, autocommit=False, fail_on_conn=True, warn_db_default=True):
+
+    ensure_required_libs(module)
```

Comment:
Would it be better to reuse the existing `ensure_libs` function here?

```suggestion
    try:
        ensure_libs(sslrootcert=module.params.get('ca_cert'))
    except LibraryError as e:
        module.fail_json(msg=str(e))
```

Reply:
@andytom , it causes problems with unit tests, see links in the description

Reply:
I have had a look at the description and the [failed tests in shippable](https://app.shippable.com/github/ansible/ansible/runs/120738/5/console), my understanding for checking that is that this is because the `ensure_libs` function was changed but the tests were not updated. What I am suggesting above is not to change the `ensure_libs` function but to simply use it. This will mean that we won't have two functions `ensure_libs` and `ensure_required_libs` that contain the same logic.

Reply:
> ensure_libs function was changed but the tests were not updated

Exactly, my suggestion from ansible-devel:
```
I examined unit tests... They cover only ensure_libs (small strange function) and they must be completely rewritten.
I made a note about it in our group. However I can't implement they fast because didn't work with unittest and other at all.
If I start to learn unittest, etc. now, I'm afraid it takes significant time (weeks maybe, I'm working on my day job at the same time).
Of course I do want to figure it out because this is a very important skill for a developer but now this changes (moving params to connect_to_db) is a blocker for me for making other PRs
So I suggest this PR https://github.com/ansible/ansible/pull/55799.
Nothing criminal, in short, I left ensure_libs() for a while and added new one instead.
Better I'll understand pytest lib completely and add unit test by separate special PR later (I made the note in the working group, so I won't forget).
Now I suggest to merge this PR (I copied files from the previous that was merged, only changed postgres.py and postgresql_db a bit)
```
In short it needs a lot of time but this PR is a blocker. Sure we'll remove one of them later.

Reply:
So the suggestion I added above (copied below) works around this, if we add the following change we don't need to add the `ensure_required_libs` function or make any changes to  `ensure_libs`.

```suggestion
    try:
        ensure_libs(sslrootcert=module.params.get('ca_cert'))
    except LibraryError as e:
        module.fail_json(msg=str(e))
```

Reply:
Have you run anything like this after this changes?
```
ansible-test units test/units/module_utils/test_postgresql.py --python 2.7 --docker fedora28 -vvvv > ~/test.log
```

Reply:
I'll try it on Monday (hope), thank you!

Reply:
if psycopg2 is not installed, it returns:
```
 The full traceback is:
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_postgresql_query_payload_5bN2rh/ansible_postgresql_query_payload.zip/ansible/module_utils/postgres.py", line 71, in connect_to_db
    ensure_libs(sslrootcert=module.params.get('ca_cert'))
  File "/tmp/ansible_postgresql_query_payload_5bN2rh/ansible_postgresql_query_payload.zip/ansible/module_utils/postgres.py", line 49, in ensure_libs
    raise LibraryError('psycopg2 is not installed. we need psycopg2.')

fatal: [spblnx176]: FAILED! => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    }, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "ca_cert": null, 
            "db": null, 
            "login_host": "", 
            "login_password": "", 
            "login_unix_socket": "", 
            "login_user": "postgres", 
            "named_args": null, 
            "path_to_script": null, 
            "port": 5432, 
            "positional_args": null, 
            "query": "SELECT now()", 
            "session_role": null, 
            "ssl_mode": "prefer"
        }
    }, 
    "msg": "psycopg2 is not installed. we need psycopg2."
}
```
We need to use missing_required_libs and get something like^
```
   "msg": "Failed to import the required Python library (psycopg2) on mydbhost.lan's Python /usr/bin/python. Please read module documentation and install in the appropriate location"
```
We also need to make good unit tests for functions from postgres.py from scratch using pytest.
So I suggest to leave it as it is.

Reply:
Fair enough.